pkg/router/controller/factory/factory.go

```diff
@@ -2,509 +2,295 @@ package factory
-	return r.routes.ListIndexFuncValues("host")
-}
+func (f *RouterControllerFactory) initInformers(rc *routercontroller.RouterController) {
+	f.createEndpointsSharedInformer(rc)
+	f.createRoutesSharedInformer(rc)
 
-func (r *routesByHost) Route(host string) (*routeapi.Route, bool) {
-	arr, err := r.routes.ByIndex("host", host)
-	if err != nil || len(arr) == 0 {
-		return nil, false
+	if rc.WatchNodes {
+		f.createNodesSharedInformer(rc)
 	}
-	return oldestRoute(arr), true
-}
 
-func (r *routesByHost) Endpoints(namespace, name string) *kapi.Endpoints {
-	obj, ok, err := r.endpoints.GetByKey(fmt.Sprintf("%s/%s", namespace, name))
-	if !ok || err != nil {
-		return &kapi.Endpoints{}
+	if rc.EnableIngress {
+		f.createIngressesSharedInformer(rc)
+		f.createSecretsSharedInformer(rc)
 	}
-	return obj.(*kapi.Endpoints)
-}
-
-// routeAge sorts routes from oldest to newest and is stable for all routes.
-type routeAge []routeapi.Route
-
-func (r routeAge) Len() int      { return len(r) }
-func (r routeAge) Swap(i, j int) { r[i], r[j] = r[j], r[i] }
-func (r routeAge) Less(i, j int) bool {
-	return routeapi.RouteLessThan(&r[i], &r[j])
-}
 
-func oldestRoute(routes []interface{}) *routeapi.Route {
-	var oldest *routeapi.Route
-	for i := range routes {
-		route := routes[i].(*routeapi.Route)
-		if oldest == nil || route.CreationTimestamp.Before(oldest.CreationTimestamp) {
-			oldest = route
+	// Start informers
+	for _, informer := range f.informers {
+		go informer.Run(utilwait.NeverStop)
+	}
+	// Wait for informers cache to be synced
+	for objType, informer := range f.informers {
+		if !kcache.WaitForCacheSync(utilwait.NeverStop, informer.HasSynced) {
+			utilruntime.HandleError(fmt.Errorf("failed to sync cache for %+v shared informer", objType))
 		}
 	}
-	return oldest
 }
 
-func hostIndexFunc(obj interface{}) ([]string, error) {
-	route := obj.(*routeapi.Route)
-	hosts := []string{
-		fmt.Sprintf("%s-%s%s", route.Name, route.Namespace, ".generated.local"),
+func (f *RouterControllerFactory) registerInformerEventHandlers(rc *routercontroller.RouterController) {
+	f.registerSharedInformerEventHandlers(&kapi.Endpoints{}, rc.HandleEndpoints)
+	f.registerSharedInformerEventHandlers(&routeapi.Route{}, rc.HandleRoute)
+
+	if rc.WatchNodes {
+		f.registerSharedInformerEventHandlers(&kapi.Node{}, rc.HandleNode)
 	}
-	if len(route.Spec.Host) > 0 {
-		hosts = append(hosts, route.Spec.Host)
+	if rc.EnableIngress {
+		f.registerSharedInformerEventHandlers(&extensions.Ingress{}, rc.HandleIngress)
+		f.registerSharedInformerEventHandlers(&kapi.Secret{}, rc.HandleSecret)
 	}
-	return hosts, nil
 }
 
-// routeLW is a ListWatcher for routes that can be filtered to a label, field, or
-// namespace.
-type routeLW struct {
-	client    routeclient.RoutesGetter
-	label     labels.Selector
-	field     fields.Selector
-	namespace string
+func (f *RouterControllerFactory) initCallbacks(rc *routercontroller.RouterController) {
+	rc.HasSyncedInformers = func() bool {
```

Comment:
Do you still need this?

Reply:
When re-sync is in progress, we want to reduce the number of reloads. We could fully rely on router coalescing and can get rid of this informer synced check. @knobunc @rajatchopra what do you think?

Reply:
Tested with 1000 routes to check whether router coalescing is sufficient or informer synced check is necessary to reduce reloads. Router coalescing without sync check worked fine, so removed this unnecessary check.